pilopt/src/inliner.rs

```diff
@@ -11,46 +14,48 @@ pub fn replace_constrained_witness_columns<
     T: FieldElement,
     V: Ord + Clone + Hash + Eq + Display,
 >(
-    mut constraint_system: ConstraintSystem<T, V>,
+    constraint_system: ConstraintSystem<T, V>,
     max_degree: usize,
 ) -> ConstraintSystem<T, V> {
-    loop {
-        if !try_apply_substitution(&mut constraint_system, max_degree) {
-            break;
-        }
-    }
-
-    constraint_system
-}
+    let mut to_remove_idx = HashSet::new();
+    let mut inlined_vars = HashSet::new();
 
-/// Attempts to apply one valid variable substitution across the constraint system.
-///
-/// Skips substitutions that would increase the degree beyond `max_degree`
-/// or affect variables in the `keep` set. Returns true if a substitution was applied.
-fn try_apply_substitution<T: FieldElement, V: Ord + Clone + Hash + Eq + Display>(
-    constraint_system: &mut ConstraintSystem<T, V>,
-    max_degree: usize,
-) -> bool {
-    let indices: Vec<usize> = (0..constraint_system.algebraic_constraints.len()).collect();
+    let mut constraint_system = IndexedConstraintSystem::from(constraint_system);
 
-    for idx in indices.into_iter().rev() {
-        let constraint = &constraint_system.algebraic_constraints[idx];
+    for curr_idx in (0..constraint_system.algebraic_constraints().len()).rev() {
+        let constraint = &constraint_system.algebraic_constraints()[curr_idx];
 
         for (var, expr) in find_inlinable_variables(constraint) {
-            if is_valid_substitution(&var, &expr, constraint_system, max_degree) {
+            if is_valid_substitution(&var, &expr, &constraint_system, max_degree) {
                 log::trace!("Substituting {var} = {expr}");
                 log::trace!("  (from identity {constraint})");
-                constraint_system.expressions_mut().for_each(|identity| {
-                    identity.substitute_by_unknown(&var, &expr);
-                });
 
-                constraint_system.algebraic_constraints.remove(idx);
-                return true;
+                constraint_system.substitute_by_unknown(&var, &expr);
+                to_remove_idx.insert(curr_idx);
+                inlined_vars.insert(var);
+
+                break;
             }
         }
     }
 
-    false
+    let mut constraint_system = ConstraintSystem::from(constraint_system);
+
+    // remove inlined constraints from system
+    constraint_system.algebraic_constraints = constraint_system
+        .algebraic_constraints
+        .into_iter()
+        .enumerate()
+        .filter_map(|(idx, qse)| (!to_remove_idx.contains(&idx)).then_some(qse))
+        .collect();
+
+    // sanity check
+    assert!(constraint_system.expressions_mut().all(|expr| {
```

Comment:
```suggestion
    assert!(constraint_system.expressions_mut().all(|expr| {
```

the mut is a bit surprising here, is there an immutable equivalent?